### PR TITLE
refactor: Unify update and replicated query execution algorithms

### PR DESCRIPTION
**Pull request by dsarlis**

### CODE REVIEW

File: rs/execution_environment/src/execution/call_or_task.rs [hunk 1]
```diff
@@ -522,19 +598,43 @@ impl UpdateHelper {
...
+                let result = output
+                    .wasm_result
+                    .map_err(|err| err.into_user_error(&self.canister.canister_id()));
+                let refund = match action {
+                    CallContextAction::Reply { refund, .. }
+                    | CallContextAction::Reject { refund, .. }
+                    | CallContextAction::NoResponse { refund, .. }
+                    | CallContextAction::Fail { refund, .. } => refund,
+                    CallContextAction::NotYetResponded | CallContextAction::AlreadyResponded => {
+                        original.call_or_task.cycles()
```
Comment (michael-weigelt) [inline rs/execution_environment/src/execution/call_or_task.rs:625]: What is the situation here? We are in finish, so the action `Reply` etc make sense. But what are these two latter options? And what/why do we refund exactly?

File: rs/execution_environment/src/execution/call_or_task/tests.rs [hunk 2]
```diff
@@ -115,6 +115,12 @@ fn wat_writing_to_each_stable_memory_page_query(memory_amount: u64) -> String {
     )
 }
 
+// Helper function to allow testing both update and query methods with the same test.
+fn with_update_and_replicated_query<F: FnOnce(String) + Clone>(test: F) {
+    test.clone()("update".to_string());
+    test("query".to_string());
+}
```
Comment (michael-weigelt) [inline rs/execution_environment/src/execution/call_or_task/tests.rs:122]:
```suggestion
fn with_update_and_replicated_query<F: Fn(&str)>(test: F) {
    test("update");
    test("query");
}
```
I tried it for a single test and that one worked. But because of the `String` -> `&str` change you'll have to remove a `&` in some tests.

File: rs/execution_environment/src/execution/call_or_task/tests.rs [hunk 3]
```diff
@@ -350,6 +436,94 @@ fn dts_update_concurrent_cycles_change_fails() {
...
 }
 
+#[test]
+fn dts_replicated_query_concurrent_cycles_change_fails() {
+    // Test steps:
+    // 1. Canister A starts running the query method.
+    // 2. While canister A is paused, we emulate a postponed charge
+    //    of 1000 cycles (i.e. add 1000 to `ingress_induction_cycles_debit`).
+    // 3. The query method resumes and burns 1000 cycles.
+    // 4. The query method succeeds because there are enough cycles
```
Comment (michael-weigelt) [inline rs/execution_environment/src/execution/call_or_task/tests.rs:446]:
```suggestion
    // 4. The query method fails because there are not enough cycles
```

File: rs/execution_environment/src/execution/call_or_task/tests.rs [hunk 4]
```diff
@@ -235,6 +241,86 @@ fn dts_update_concurrent_cycles_change_succeeds() {
...
+        NextExecution::None,
+    );
+
+    assert_eq!(
+        test.canister_state(canister_id).system_state.balance(),
+        initial_cycles
+            - cycles_to_burn
+            - (test.canister_execution_cost(canister_id) - initial_execution_cost)
+            - cycles_debit,
+    );
```
Comment (michael-weigelt) [inline rs/execution_environment/src/execution/call_or_task/tests.rs:321]: I suppose this check succeeding also implies that execution was successful. If there had not been enough cycles to execute the second DTS slice, then... remind me again... 1) the execution cost would be applied 2) the cycles debit would be applied 3) cycles_to_burn would NOT be applied.

File: rs/execution_environment/src/execution/call_or_task/tests.rs [hunk 5]
```diff
@@ -611,378 +785,401 @@ fn hitting_access_limit_fails_non_replicated_query() {
...
-    let mut test = ExecutionTestBuilder::new()
-        .with_instruction_limit(instruction_limit)
-        .with_slice_instruction_limit(10_000)
-        .with_manual_execution()
-        .build();
+fn dts_replicated_execution_resume_fails_due_to_call_context_change() {
+    with_update_and_replicated_query(|method| {
+        // Test steps:
+        // 1. Canister A starts running the update|query method.
+        // 2. While canister A is paused, we change its call context counter.
```
Comment (michael-weigelt) [inline rs/execution_environment/src/execution/call_or_task/tests.rs:854]: I know this test just adapts an existing test, but I don't understand its purpose. I saw the call context check in the implementation, and was wondering: How could this happen? So, how could this happen, apart from in a test?

**Review by michael-weigelt — APPROVED** [review]

File: rs/execution_environment/src/execution/call_or_task.rs [hunk 6]
```diff
@@ -158,13 +162,19 @@ pub fn execute_update(
...
             msg.method_payload().to_vec(),
             msg.cycles(),
             *msg.sender(),
             helper.call_context_id(),
         ),
+        CanisterCallOrTask::Query(msg) => ApiType::replicated_query(
+            time,
+            msg.method_payload().to_vec(),
+            *msg.sender(),
+            helper.call_context_id(),
```
[the remark this answers is not in the document]
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task.rs:176]: Good point, I will rename to `CallOrTaskHelper` following the naming convention I decided on.

File: rs/execution_environment/src/execution/call_or_task.rs (hunk 1 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task.rs:625]: See the [discussion](https://github.com/dfinity/ic/pull/363/files/9f7991d4058aa32b647a8ae16420538f7c05cb7b#r1923334159) in the previous PR that introduced this.

File: rs/execution_environment/src/execution/call_or_task/tests.rs (hunk 2 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task/tests.rs:122]: I can live with removing `&`, thanks for the suggestion.

File: rs/execution_environment/src/execution/call_or_task/tests.rs (hunk 4 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task/tests.rs:321]: Yes correct.

File: rs/execution_environment/src/execution/call_or_task/tests.rs (hunk 5 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task/tests.rs:854]: Bug in the code? I think the test is here to check an edge case given that the DTS helper requires to find a call context with a specific id so it can work properly. Under a correct implementation it would indeed not fail.

Comment (michael-weigelt) [reply rs/execution_environment/src/execution/call_or_task/tests.rs:854]: That makes sense. 

Perhaps I am misremembering, but I thought I saw a `UserError` being returned which I took as a hint that it's not just about bugs.

File: rs/execution_environment/src/execution/call_or_task/tests.rs (hunk 2 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task/tests.rs:122]: Wait, you misunderstood the problem here. The clone is on the function, not the string. The compiler somehow doesn't allow me to reuse the test function as it considers it moved once I call it once.

File: rs/execution_environment/src/execution/call_or_task/tests.rs (hunk 5 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/call_or_task/tests.rs:854]: It's a `UserError` because we cannot apply state changes but the underlying cause is clearly a indication of a bug (`FailedToApplyStateChanges` could only happen when bugs occur). And we decide to not crash but simply pretend execution failed and let the user retry.

File: rs/types/types/src/messages.rs [hunk 7]
```diff
@@ -520,21 +520,22 @@ impl Display for CanisterMessageOrTask {
 /// A wrapper around canister messages and tasks.
 #[derive(Clone, Eq, PartialEq, Hash, Debug)]
 pub enum CanisterCallOrTask {
-    Call(CanisterCall),
+    Update(CanisterCall),
+    Query(CanisterCall),
     Task(CanisterTask),
 }
```
Comment (alin-at-dfinity) [inline rs/types/types/src/messages.rs:526]: It may be worth mentioning somewhere in here that these are replicated executions (i.e. `Query` never stands for a non-replicated query.

**Review by alin-at-dfinity — APPROVED** [review]

File: rs/types/types/src/messages.rs (hunk 7 above)
Comment (dsarlis) [reply rs/types/types/src/messages.rs:526]: I can enhance the doc comment. The name of the enum (although convoluted) should capture it. `call` is usually meant to be something executed in replicated mode.
